### PR TITLE
Require representation input to be Quantity, and backward-compatibility layer

### PR DESCRIPTION
**Pull request by astrofrog**

### CODE REVIEW

**Comment by astrofrog** [comment]
@eteq - this is almost completely functional, but there are a few test failures. Before I fix them, I just wanted to check whether it's ok for frames to have positional arguments for the coordinates?

**Comment by astrofrog** [comment]
I'm working on the failures under the assumption that positional arguments are ok (based on the API document)

**Comment by astrofrog** [comment]
So just to be clear, this PR means that:
- Representations have to take `Quantity`, nothing else (nicer & simpler behavior)
- Frames then by extension can't take anything other than `Quantity`
- For 0.4, if frames detect a deprecated API use, they will emit a deprecation warning and return a `SkyCoord` instance.

I believe this actually provides a mostly backward-compatibility layer for 0.4. We could actually try and run the old test suite on it to check.

**Comment by astrofrog** [comment]
All current tests pass! A few documentation (rst) tests fail, but the documentation needs updating anyway.

**Comment by astrofrog** [comment]
Also, one nice thing here is that the backward-compatibility layer is easy to remove in the next version - we just need to remove `__new__` from the base frame class.

File: astropy/coordinates/tests/test_api_ape5.py [hunk 1]
```diff
@@ -59,10 +59,6 @@ def test_representations_api():
     # Default is to copy arrays, but optionally, it can be a reference
     UnitSphericalRepresentation(lon=[8, 9]*u.hourangle, lat=[5, 6]*u.deg, copy=False)
 
-    # strings are parsed by `Latitude` and `Longitude` constructors, so no need to
-    # implement parsing in the Representation classes
-    UnitSphericalRepresentation(lon='2h6m3.3s', lat='0.1rad')
```
Comment (eteq) [inline astropy/coordinates/tests/test_api_ape5.py:64]: actually, can you replace this with something like:

```
UnitSphericalRepresentation(lon=Angle('2h6m3.3s'), lat=Angle('0.1rad'))
```

just so we make sure this use case is addressed?

File: astropy/coordinates/tests/test_api_ape5.py [hunk 2]
```diff
@@ -80,14 +76,6 @@ def test_representations_api():
     with raises(ValueError):
         c2 = UnitSphericalRepresentation(lon=[8, 9, 10]*u.hourangle, lat=[5, 6]*u.deg)
 
-    # It's also possible to pass in scalar quantity lists with mixed units. These
-    # are converted to array quantities following the same rule as `Quantity`: all
-    # elements are converted to match the first element's units.
-    c2 = UnitSphericalRepresentation(lon=[8*u.hourangle, 135*u.deg],
-                                        lat=[5*u.deg, (6*np.pi/180)*u.rad])
-    assert c2.lat.unit == u.deg and c2.lon.unit == u.hourangle
-    npt.assert_almost_equal(c2.lon[1].value, 9)
```
Comment (eteq) [inline astropy/coordinates/tests/test_api_ape5.py:89]: same as above (wrap in `Quantity` or `Angle` constructor), mostly just to make sure all the little conversions work as expected?

File: astropy/coordinates/baseframe.py [hunk 3]
```diff
@@ -131,6 +133,53 @@ class BaseCoordinateFrame(object):
     frame_attr_names = {}  # maps attribute to default value
     time_attr_names = ('equinox', 'obstime')  # Attributes that must be Time objects
 
+    def __new__(cls, *args, **kwargs):
```
Comment (eteq) [inline astropy/coordinates/baseframe.py:139]: maybe just add a comment here that says `TODO: remove the __new__ method in v1.0 to remove the compatibility layer` just so we don't forget that it's that easy?

**Comment by eteq** [comment]
Looks generally good, @astrofrog (a few minor comments inline)!  But it looks like you did _not_ change `CartesianRepresenation` to only accept quantities - is that intentional?

**Comment by astrofrog** [comment]
@eteq - I can change `CartesianRepresentation` too - I just hadn't touched it since it doesn't accept strings anyway, but I agree let's go all the way and make it consistent.

**Comment by eteq** [comment]
Ahh, I see.  Yeah, sounds good.

File: astropy/coordinates/tests/test_api_ape5.py (hunk 1 above)
Comment (astrofrog) [inline astropy/coordinates/tests/test_api_ape5.py:64]: Done

File: astropy/coordinates/tests/test_api_ape5.py (hunk 2 above)
Comment (astrofrog) [inline astropy/coordinates/tests/test_api_ape5.py:89]: Done

File: astropy/coordinates/baseframe.py (hunk 3 above)
Comment (astrofrog) [inline astropy/coordinates/baseframe.py:139]: Done

**Comment by astrofrog** [comment]
@eteq - hmm, it seems that the code was relying on being able to pass non-quantities to the cartesian class, e.g.:

```
        #now re-cast into an appropriate Representation, and precess if need be
        if isinstance(fk4noecoord.data, UnitSphericalRepresentation):
>           representation = CartesianRepresentation(x=x, y=y, z=z)

astropy/coordinates/builtin_frames.py:471
```

Maybe I should revert this change for now? I'm getting 160+ failures now...

**Comment by astrofrog** [comment]
In the above case, `x`, `y`, and `z` are plain numpy arrays.

**Comment by astrofrog** [comment]
Actually I think the fix is easy...

**Comment by astrofrog** [comment]
@eteq - this is ready for review/merging. All tests pass (except a few in the docs, which is to be expected)

**Comment by eteq** [comment]
Cool, thanks.  Merging!

**Pull request merged**